File: boj/GraphTheory/BOJ_14940.cpp
/*
 21.10.19
 백준 14940 쉬운 최단거리
 https://github.com/skyqnaqna/algorithm_study
 */

#include <bits/stdc++.h>

using namespace std;
using ll = long long;
using dd = double;
using pii = pair<int, int>;
using pll = pair<ll, ll>;

#define INF (int)2e9
#define endl "\n"
#define TXTOPEN freopen("input.txt", "r", stdin)

int n, m, s, e, graph[1000][1000];
vector <vector <int> > dist (1000, vector <int> (1000, -1));
int dy[] = {-1, 1, 0, 0};
int dx[] = {0, 0, -1, 1};

bool inBound(int r, int c) {
    return 0 <= r && r < n && 0 <= c && c < m;
}

int main() {
    ios::sync_with_stdio(0); cin.tie(0);

    cin >> n >> m;

    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < m; ++j) {
            cin >> graph[i][j];
            if (graph[i][j] == 2) {
                s = i;
                e = j;
            } else if (graph[i][j] == 0) dist[i][j] = 0;
        }
    }

    dist[s][e] = 0;

    queue <pii> q;
    q.push({s, e});
    while (!q.empty()) {
        int r = q.front().first;
        int c = q.front().second;
        q.pop();

        for (int i = 0; i < 4; ++i) {
            int nr = r + dy[i];
            int nc = c + dx[i];

            if (inBound(nr, nc) && graph[nr][nc] > 0 && dist[nr][nc] == -1) {
                dist[nr][nc] = dist[r][c] + 1;
                q.push({nr, nc});
            }
        }
    }

    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < m; ++j) {
            cout << dist[i][j] << " ";
        }
        cout << endl;
    }

    return 0;
}

File: boj/GraphTheory/BOJ_14940.swift
/*
 백준 14940 쉬운 최단거리
 21.11.16
 https://github.com/skyqnaqna/algorithm_study
 */

import Foundation

let direction = [(-1, 0), (1, 0), (0, -1), (0, 1)]
let input = readLine()!.split(separator: " ").map { Int(String($0))! }
let n = input[0], m = input[1]

var graph = [[Int]](repeating: [Int](repeating: -1, count: m), count: n)
var start = (0, 0)

for i in 0 ..< n {
  let row = readLine()!.split(separator: " ").map { Int(String($0))! }
  
  for j in 0 ..< m {
    if row[j] == 0 {
      graph[i][j] = 0
    }
    if row[j] == 2 {
      start = (i, j)
      graph[i][j] = 0
    }
  }
}

func inBound(_ r: Int, _ c: Int) -> Bool {
  return 0 <= r && r < n && 0 <= c && c < m
}

var q = [(Int, Int, Int)](), front = 0

q.append((start.0, start.1, 0))

while front < q.count {
  let (r, c, w) = q[front]
  front += 1
  
  for (dy, dx) in direction {
    let nr = r + dy
    let nc = c + dx
    
    if !inBound(nr, nc) { continue }
    
    if graph[nr][nc] == -1 {
      graph[nr][nc] = w + 1
      q.append((nr, nc, w + 1))
    }
  }
}

var answer = ""
for i in 0 ..< n {
  answer += graph[i].map { String($0) }.joined(separator: " ") + "\n"
}

print(answer)
